### DISCUSSION TITLE
ESP32 WROVER and PSRAM

### DISCUSSION THREAD

**Discussion by augea** [arendst/Tasmota#16423] [Q&A] [answered]
I bought an ESP32 WROVER module and I was able to compile Tasmota V12.
To get PSRAM working I uses the build_flags in platformio_override.ini. 
I was able to set CPU frequency, but flags 
-DBOARD_HAS_PSRAM
-mfix-esp32-psram-cache-issue
did not work.
After some test I found that if I change the board from ESP32-4M to ESP32-cam PSRAM was available. I checked the board file and saw the necessary compile flags, too.
After that I changed the esp32_4M board file and added the compile flags there.
In this case also PSRAM worked.
So my question is, to get the WROVER module working, do I have to create a new board file since there are a lot of valid combinations : (4M, 8M, 16M flash, with or without PSRAM, with or without PSRAM issue (newer CPU don’t need the fix), CPU frequencys.)
Is this correct to create additional board files or is there a different way to configure the necessary combination. In my tests I used the latest Tasmota stable version with platformio on Visual Studio Code.
Do I have to create my own board file?

**Comment by s-hadinger** [comment, marked as answer]
Why don't you use the latest Tasmota version? PSRAM is already included

**Reply by augea** [reply]
I did, it is V12. Sorry for the mistake. I correct this. Exact version is: 12.1.1
I have downloaded the latest version and compiled "tasmota32" build by uncommenting in platformio_override.ini.

Then I added the additional bulid_flags but I seems they are overwritten with the content of the board file. If I mistype for example -DBOARD_HAS_PSRAM the compiler reports an error. If typed correct the information (web page) of the Tasmota does not report PSRAM. Only if I change the board file or edit the board file that has the PSRAM build parameters inside, PSRAM is used.

**Reply by s-hadinger** [reply]
What does the report page shows? Also set `weblog 4` and report the first lines of the boot. 

Special flags are needed for early revisions of Esp32 but they are not necessary unless you bought a very old module

**Reply by sfromis** [reply]
What works for me on a board with an early revision `ESP32-D0WDQ6 rev.1` is this build environment in `platformio_tasmota_cenv.ini`:
```
[env:tasmota32_psram]
extends                 = env:tasmota32
build_flags             = ${env:tasmota32_base.build_flags} -DBOARD_HAS_PSRAM -DHAS_PSRAM_FIX -mfix-esp32-psram-cache-issue -mfix-esp32-psram-cache-strategy=memw
```

Those *crappy* ESP32-CAM boards tend to have that early revision, hence the fixing is included in that build.

**Reply by augea** [reply]
It is an early revision and requires a fix. It is not a cam board but I used it to get PSRAM active. It is just an older WROVER module I ordered on Aliexpress. This is the result with cam Board.

ESP Chip Id	13424932 (ESP32-D0WDQ6 rev.1)
Flash Chip Id	0x16405E
Flash Size	4096 KB
Program Flash Size	4096 KB
Program Size	1448 KB
Free Program Space	1431 KB
Free Memory	158.8 KB (frag. 31%)
PS-RAM Memory	4093 KB
PS-RAM free Memory	4088 KB
 
Partition safeboot	832 KB (used 91%)
Partition app0*	2880 KB (used 50%)
Partition fs	320 KB

If I compile with tasmota32 with this "platformio_override.ini" 

`; ***  Example PlatformIO Project Configuration Override File   ***
; ***  Changes done here override settings in platformio.ini    ***
;
; *****************************************************************
; ***  to activate rename this file to platformio_override.ini  ***
; *****************************************************************
;
; Please visit documentation for the options and examples
; http://docs.platformio.org/en/stable/projectconf.html

[platformio]
; For best Gitpod performance remove the ";" in the next line. Needed Platformio files are cached and installed at first run
;core_dir = .platformio
; For unrelated compile errors with Windows it can help to shorten Platformio project path
;workspace_dir = c:\.pio

; *** Build/upload environment
default_envs            =
; *** Uncomment the line(s) below to select version(s)
;                          tasmota
;                          tasmota-debug
;                          tasmota-minimal
;                          tasmota-lite
;                          tasmota-knx
;                          tasmota-sensors
;                          tasmota-display
;                          tasmota-zbbridge
;                          tasmota-ir
                          **tasmota32**
;                          tasmota32-zbbrdgpro
;                          tasmota32-bluetooth
;                          tasmota32-webcam
;                          tasmota32-knx
;                          tasmota32-sensors
;                          tasmota32-lvgl
;                          tasmota32-ir
;                          tasmota32solo1
;                          tasmota32c3
;                          tasmota32c3cdc
;                          tasmota32s2
;                          tasmota32s2cdc
;                          tasmota32s3
;                          tasmota32s3cdc
;                          tasmota32-odroidgo
;                          tasmota32-core2
;                          tasmota32-nspanel


[env]
; Activate Development core by removing ";" the next lines
;platform                = https://github.com/platformio/platform-espressif8266.git
;platform_packages       = framework-arduinoespressif8266 @ https://github.com/esp8266/Arduino.git
;                          mcspr/toolchain-xtensa @ ~5.100300.211127
;                          platformio/tool-esptoolpy @ ~1.30300
;build_unflags           = ${common.build_unflags}
;                          -Wswitch-unreachable
;build_flags             = ${common.build_flags}
;                          -D PIO_FRAMEWORK_ARDUINO_MMU_CACHE16_IRAM48_SECHEAP_SHARED
;                          -Wno-switch-unreachable
; *** Optional Debug messages
;                         -DDEBUG_TASMOTA_CORE
;                         -DDEBUG_TASMOTA_DRIVER
;                         -DDEBUG_TASMOTA_SENSOR
; Build variant 1MB = 1MB firmware no filesystem (default)
board                   = ${common.board}
; Build variant 2MB = 1MB firmware, 1MB filesystem (most Shelly devices)
;board                   = esp8266_2M1M
; Build variant 4MB = 1MB firmware, 1MB OTA, 2MB filesystem (WEMOS D1 Mini, NodeMCU, Sonoff POW)
;board                   = esp8266_4M2M
;board_build.f_cpu       = 160000000L
;board_build.f_flash     = 40000000L
;monitor_speed           = 115200
; *** Serial port used for erasing/flashing the ESP82xx
;upload_port             = COM5
extra_scripts           = ${esp_defaults.extra_scripts}
;                          pio-tools/obj-dump.py
lib_ignore              =
                          Servo(esp8266)
                          ESP8266AVRISP
                          ESP8266LLMNR
                          ESP8266NetBIOS
                          ESP8266SSDP
                          SP8266WiFiMesh
                          Ethernet(esp8266)
                          GDBStub
                          TFT_Touch_Shield_V2
                          ESP8266HTTPUpdateServer
                          ESP8266WiFiMesh
                          EspSoftwareSerial
                          SPISlave
                          Hash
; Disable next if you want to use ArduinoOTA in Tasmota (default disabled)
                          ArduinoOTA
lib_extra_dirs          = ${library.lib_extra_dirs}


[env:tasmota32_base]
; *** Uncomment next lines ";" to enable development Tasmota Arduino version ESP32
;platform                = https://github.com/tasmota/platform-espressif32/releases/download/v2.0.4.1/platform-espressif32-2.0.4.1.zip
;platform_packages       = framework-arduinoespressif32 @ https://github.com/Jason2866/esp32-arduino-lib-builder/releases/download/927/framework-arduinoespressif32-442_esp41-13eca437d1.tar.gz
build_unflags           = ${esp32_defaults.build_unflags}
build_flags             = ${esp32_defaults.build_flags} **-DBOARD_HAS_PSRAM -mfix-esp32-psram-cache-issue**
; Build variant ESP32 4M Flash, Tasmota 1856k Code/OTA, 1344k LITTLEFS (default)
**board                   = esp32_4M**
; Build variant ESP32 8M Flash, Tasmota 2944k Code/OTA, 2112k LITTLEFS
;board                   = esp32_8M
; Build variant ESP32 16M Flash, Tasmota 2944k Code/OTA, 10M LITTLEFS
;board                   = esp32_16M
**board_build.f_cpu       = 240000000L**
;board_build.f_flash     = 40000000L
monitor_speed           = 115200
; *** Serial port used for erasing/flashing the ESP32
;upload_port             = ${common.upload_port}
upload_port             = COM3
;upload_speed            = 115200
upload_resetmethod      = ${common.upload_resetmethod}
extra_scripts           = ${esp32_defaults.extra_scripts}
;                          pio-tools/obj-dump.py
lib_ignore              =
                          HTTPUpdateServer
                          ESP RainMaker
                          WiFiProv
                          USB
                          SPIFFS
                          ESP32 Azure IoT Arduino
                          ESP32 Async UDP
                          ESP32 BLE Arduino
;                          SimpleBLE
                          NetBIOS
                          ESP32
                          Preferences
                          BluetoothSerial
; Disable next if you want to use ArduinoOTA in Tasmota32 (default disabled)
                          ArduinoOTA

lib_extra_dirs           = ${library.lib_extra_dirs}
; *** ESP32 lib. ALWAYS needed for ESP32 !!!
                          lib/libesp32
; *** comment the following line if you dont use LVGL in a Tasmota32 build. Reduces compile time
                          lib/libesp32_lvgl
; *** comment the following line if you dont use ESP32 Audio in a Tasmota32 build. Reduces compile time
;                          lib/libesp32_audio
; *** uncomment the following line if you use Bluetooth or Apple Homekit in a Tasmota32 build. Reduces compile time
;                          lib/libesp32_div
; *** uncomment the following line if you use Epaper driver epidy in your Tasmota32 build. Reduces compile time
;                          lib/libesp32_eink


[library]
shared_libdeps_dir      = lib
; *** Library disable / enable for variant Tasmota(32). Disable reduces compile time
; *** !!! Disabling needed libs will generate compile errors !!!
; *** The resulting firmware will NOT be different if you leave all libs enabled
; *** Disabling by putting a ";" in front of the lib name
; *** If you dont know what it is all about, do not change
lib_extra_dirs           =
; *** Only disabled for Tasmota minimal and Tasmota light. For all other variants needed!
                           lib/lib_basic
; **** I2C devices. Most sensors. Disable only if you dont have ANY I2C device enabled
                           lib/lib_i2c
; *** Displays. Disable if you dont have any Display activated
                           lib/lib_display
; *** Bear SSL and base64. Disable if you dont have SSL or TLS activated
                           lib/lib_ssl
; *** Audio needs a lot of time to compile. Mostly not used functions. Recommended to disable
                           lib/lib_audio
; *** RF 433 stuff (not RF Bridge). Recommended to disable
                           lib/lib_rf
; *** Mostly not used functions. Recommended to disable
                           lib/lib_div
`
I get this: (It is the same as result as with the pre-compiled version "tasmaota32 buetooth") That's why I tried to compile my own version. In the end I need my own compilation because I need to mix IR, sensors and displays in one software.

00:00:00.002-212/50 HDW: **ESP32-D0WDQ6 (PSRAM disabled)**
00:00:00.003-212/50 HDW: PSRAM is disabled, requires specific compilation on this hardware (see doc)
00:00:00.048-210/49 UFS: FlashFS mounted with 304 kB free
00:00:00.090 CFG: Loaded from File, Count 28

I did not use "platformio_tasmota_cenv.ini"because I need just one version.

**Reply by sfromis** [reply]
You did not try with the build_flags I found to be working?

**Reply by augea** [reply]
@sfromis Your bulid_flags work. I did not expect that since the flags set before were the official flags recommend by Espressif.

**Problem is solved with** 
      build_flags **-DBOARD_HAS_PSRAM -DHAS_PSRAM_FIX -mfix-esp32-psram-cache-issue -mfix-esp32-psram-cache-strategy=memw** 
for ESP32 WROVER (ESP32-D0WDQ6 rev.1)

Thank you for your support!

**Reply by Jason2866** [reply]
> @sfromis Your bulid_flags work. I did not expect that since the flags set before were the official flags recommend by Espressif.
> 
> **Problem is solved with** build_flags **-DBOARD_HAS_PSRAM -DHAS_PSRAM_FIX -mfix-esp32-psram-cache-issue -mfix-esp32-psram-cache-strategy=memw** for ESP32 WROVER (ESP32-D0WDQ6 rev.1)
> 
> Thank you for your support!

We have added functions in Tasmota to activate PSRAM for devices which do not need this patches (most ones). In the logs of your device is a message, which does tell the device needs the patches for working PSRAM. Tasmota needs a additional flag for this devices which tells the bug fix is compiled in in the binary. There is no possibility to detect this bug fix in Tasmota

**Reply by augea** [reply]
Checked the hardware of the module. 
It has a 64Mbit PSRAM chip inside. (IPS6404) 
Tasmota should report 8MB PSRAM but it only reports 4MB. 
Do I need additional build_flags?

Berry tasmota.memory () reports this:
{'stack_low': 3, 'program_free': 1431, 'program': 1448, '**psram': 4093**, 'stack_size': 8, 'flash_real': 4096, 'iram_free': 29, 'flash': 4096, 'frag': 32, 'heap_free': 156, 'psram_free': 4086}

The information page is similar:

**PS-RAM Memory 4093 KB**
PS-RAM free Memory 4082 KB

**Reply by s-hadinger** [reply]
There is a hardware limitation in ESP32 chip: only 4MB of PSRAM can be mapped to Memory. The additional memory needs paging, which neither Arduino nor Tasmota support. We haven't found any meaningful use for the paging mechanism

**Reply by augea** [reply]
@s-hadinger Thank you for your feedback. Can live with 4MB. I just was wondering.

**Reply by Silverstar** [reply]
> Why don't you use the latest Tasmota version? PSRAM is already included

I'll continue this thread instead of opening a new one; I just obtained a Lanbon L8 (WROVER-B) and flashed the latest development (12.4.0.2) tasmota32-lvgl.

ESP Chip Id | 7335560 (ESP32-D0WD rev.1)
-- | --
Flash Chip Id | 0x1740C8 (DIO)

In the console, I received the
`PSRAM is disabled, requires specific compilation on this hardware (see doc)`
message upon boot as well. I found this thread, got through vs code platform.io and added the build flags to the existing ones for tasmota32-lvgl and now I have psram available.

> `-DBOARD_HAS_PSRAM -DHAS_PSRAM_FIX -mfix-esp32-psram-cache-issue -mfix-esp32-psram-cache-strategy=memw`

PS-RAM Memory | 4091 KB
-- | --
PS-RAM free Memory | 4040 KB

That part I'm capable by myself, but still, I would like to use the precompiled binaries instead of compiling it myself everytime and others may take advance of this as well.
Is it possible to add such build flags by a define (there is USE_LANBON_L8 already)?
So at least it could be used by that online click what you need and add user defines here - compiler site?
I'm fine to provide a PR if I would know what to change. If it could be included in every tasmota32 build without side effects, I would have been done I guess. Or is there another way?

**Reply by Jason2866** [reply]
The files are needed ONLY for esp32 rev1 with connected PSRAM. Later revisions do not need this (fix) files.
We do not add to all since it wastes flash space. For most it is not needed.
Just use board `esp32-fix` instead of `esp32` and the PSRAM fix will be included when compiling. No compile switches needed
A pity when new devices are still sold with this old esp32 chips (not produced since years...)

**Reply by Silverstar** [reply]
I see. Thank you for elaboration!

**Reply by GasTurbineMan** [reply]
> The files are needed ONLY for esp32 rev1 with connected PSRAM. Later revisions do not need this (fix) files. We do not add to all since it wastes flash space. For most it is not needed. Just use board `esp32-fix` instead of `esp32` and the PSRAM fix will be included when compiling. No compile switches needed A pity when new devices are still sold with this old esp32 chips (not produced since years...)

Here we are in July 2025 and rev1 boards still being sold.  Not knowing they were rev1, I just purchased some for a "great" price.  Luckily, I found this post.  "esp32-fix" still works.

Thanks
